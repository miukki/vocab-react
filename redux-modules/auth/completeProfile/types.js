const prefix = 'auth/completeProfile'

const types = {}
types.set = function(type) {
  this[type] = `${prefix}/${type}`
}

types.set('SHOW')
types.set('HIDE')

types.set('CHANGE_NAME')
types.set('SUBMIT')
types.set('SUBMIT_DONE')
types.set('SHOW_ERROR')
types.set('DISMISS_ERROR')
types.set('RESET_STATE')

export default types
