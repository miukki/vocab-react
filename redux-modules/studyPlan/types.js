const prefix = 'eic-vocab-app/studyPlan'

const types = {}
types.set = function(type) {
  this[type] = `${prefix}/${type}`
}

// picker
types.set('UPDATE_PICKER_DAILY_WORDS_AMOUNT')
types.set('UPDATE_PICKER_DAYS_AMOUNT')

// user data
types.set('LOAD_SAVED_DATA')
types.set('LOAD_SAVED_DATA_DONE')

// change user data
types.set('CHANGE_WORDS_LIST')
types.set('UPDATE_WORDS_LIST')

types.set('CONFIRM_PICKER')
types.set('UPDATE_DAILY_WORDS_AMOUNT')

types.set('ADD_LEARNED_WORD')
types.set('UPDATE_LEARNED_WORDS')

types.set('ADD_CHECKIN_LOG')
types.set('UPDATE_CHECKIN_LOGS')

types.set('UPDATE_USER_DATA') // for reducer update user data state

// save user data
types.set('SAVE_USER_DATA_DONE')

// others
types.set('UPDATE_CURRENT_DATE')
types.set('CHECK_DATE')

export default types
