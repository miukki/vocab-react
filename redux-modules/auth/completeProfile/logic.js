import { createLogic } from 'redux-logic'
import CustomerIamApiSdkJs from 'customer-iam-api-sdk-js'
import { NavigationActions } from 'react-navigation'
import typesSignUp from '../signUp/types'
import types from './types'
import Validator from '../../../lib/Validator'

const getJsState = function(getState) {
  return getState().toJS().completeProfile
}
const getJsStateToken = function(getState) {
  return getState().toJS().token
}

const showCompleteProfileAfterSignUp = createLogic({
  type: typesSignUp.SUBMIT_DONE,
  process({ getState, action }, dispatch, done) {
    dispatch({ type: typesSignUp.HIDE })
    dispatch({ type: types.SHOW })
    done()
  }
})

const validate = createLogic({
  type: [types.SUBMIT],
  validate({ getState, action }, allow, reject, done) {
    const state = getJsState(getState)
    let errorTypes = []
    if (!Validator.isNameCorrect(state.name)) {
      errorTypes.push({ errorType: 'wrongName' })
    }
    if (errorTypes.length) {
      reject({ type: types.SHOW_ERROR, payload: { errorTypes } })
    }
    allow(action)
  }
})

const submitProfile = createLogic({
  type: types.SUBMIT,
  process({ customerIamApiSdkJs, Storage, getState, action }, dispatch, done) {
    //call API
    customerIamApiSdkJs
      .patchUser(getJsState(getState).name, getJsStateToken(getState).jwt)
      .then(res => {
        dispatch({
          type: types.SUBMIT_DONE
        })
        dispatch(NavigationActions.navigate({ routeName: 'Tab' }))
        done()
      })
      .catch(error => {
        if (error === CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'networkError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.SERVER_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'serverError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.USER_PASSWORD_DISMATCH) {
          let errorTypes = []
          errorTypes.push({ errorType: 'userPasswordMismatch' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        let errorTypes = []
        errorTypes.push({ errorType: 'otherError' })
        dispatch({
          type: types.SHOW_ERROR,
          payload: { errorTypes }
        })
        done()
      })
  }
})

export default [validate, showCompleteProfileAfterSignUp, submitProfile]
