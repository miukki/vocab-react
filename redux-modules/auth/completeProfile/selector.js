import Validator from '../../../lib/Validator'
// status in boolean

export const selectShow = state => {
  return state.toJS().show
}

export const selectSubmitting = state => {
  return state.toJS().submitting
}

export const selectHasError = state => {
  return Validator.isAnyValidateError(state.toJS().error)
}

// values
export const selectName = state => {
  return state.toJS().name
}

export const selectErrorTypes = state => {
  return Validator.getErrorTypes(state)
}
