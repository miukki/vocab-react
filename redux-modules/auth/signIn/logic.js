import { createLogic } from 'redux-logic'
import CustomerIamApiSdkJs from 'customer-iam-api-sdk-js'
import { NavigationActions } from 'react-navigation'
import * as tokenActions from '../token/actions'
import types from './types'
import Validator from '../../../lib/Validator'

const getJsState = function(getState) {
  return getState().toJS().signIn
}

const validate = createLogic({
  type: [types.SUBMIT],
  validate({ getState, action }, allow, reject, done) {
    const state = getJsState(getState)
    let errorTypes = []
    if (!Validator.isMobileNumberCorrect(state.mobile)) {
      errorTypes.push({ errorType: 'wrongMobile' })
    }
    if (errorTypes.length) {
      reject({ type: types.SHOW_ERROR, payload: { errorTypes } })
    }
    allow(action)
  }
})

const requestJWT = createLogic({
  type: types.SUBMIT,
  process({ customerIamApiSdkJs, Storage, getState, action }, dispatch, done) {
    customerIamApiSdkJs
      .signIn(getJsState(getState).mobile, getJsState(getState).password)
      .then(res => {
        Storage.setJWT(res.data.token).then(() => {
          dispatch({
            type: types.SUBMIT_DONE
          })

          dispatch(tokenActions.setJWT(res.data.token))
          dispatch(NavigationActions.navigate({ routeName: 'Tab' }))
          done()
        })
      })
      .catch(error => {
        // setTimeout(() => {
        if (error === CustomerIamApiSdkJs.ERRORS.NETWORK_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'networkError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.SERVER_ERROR) {
          let errorTypes = []
          errorTypes.push({ errorType: 'serverError' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        if (error === CustomerIamApiSdkJs.ERRORS.USER_PASSWORD_DISMATCH) {
          let errorTypes = []
          errorTypes.push({ errorType: 'userPasswordMismatch' })
          dispatch({
            type: types.SHOW_ERROR,
            payload: { errorTypes }
          })
          return done()
        }

        let errorTypes = []
        errorTypes.push({ errorType: 'otherError' })
        dispatch({
          type: types.SHOW_ERROR,
          payload: { errorTypes }
        })
        done()
        // }, 3000)
      })
  }
})

export default [validate, requestJWT]
