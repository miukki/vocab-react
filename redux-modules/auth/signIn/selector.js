import Validator from '../../../lib/Validator'

// status in boolean
export const selectShow = state => {
  return state.toJS().show
}

export const selectSubmitting = state => {
  return state.toJS().submitting
}

export const selectShowPassword = state => {
  return state.toJS().showPassword
}

export const selectHasError = state => {
  return Validator.isAnyValidateError(state.toJS().error)
}

// values
export const selectMobile = state => {
  return state.toJS().mobile
}

export const selectPassword = state => {
  return state.toJS().password
}

export const selectErrorTypes = state => {
  return Validator.getErrorTypes(state)
}
