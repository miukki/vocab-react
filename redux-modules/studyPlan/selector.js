import models from 'eic-vocab-app-models'
import wordsLists, { all as allWords } from '../../data/words_lists'

export const selectOpts = state => {
  let { learnedWords } = state.toJS()
  const wordsList = selectWordsList(state)

  if (!wordsList) {
    return null
  }

  const wordsAmount = models.StudyPlan.getWordsAmount(
    wordsList,
    Array.from(learnedWords)
  )

  return models.StudyPlan.getOpts(wordsAmount)
}

export const selectWordsList = state => {
  const wordsListUid = selectWordsListUid(state)
  return Array.from(wordsLists.get(wordsListUid))
}

// picker choices data

export const selectDailyWordsAmounts = state =>
  Array.from(selectOpts(state).keys())

export const selectDaysAmounts = state =>
  selectOpts(state)
    ? Array.from(new Set(selectOpts(state).values()).add(1))
    : []

// picker selected value

export const selectPickerDailyWordsAmount = state =>
  state.toJS().picker.dailyWordsAmount

export const selectPickerDaysAmount = state => {
  const opts = selectOpts(state)
  const dailyWordsAmount = selectPickerDailyWordsAmount(state)
  if (!opts) {
    return NaN
  }
  return models.StudyPlan.getDaysAmount(selectOpts(state), dailyWordsAmount)
}

// panel info

export const selectDailyWordsAmount = state => {
  return (
    state.toJS().dailyWordsAmountOfWordsList[selectWordsListUid(state)] || 15
  )
}

export const selectWordsListUid = state => {
  let { wordsListUid } = state.toJS()

  if (Array.from(wordsLists.keys()).indexOf(wordsListUid) === -1) {
    return Array.from(wordsLists.keys())[0] // get first by default
  }

  return wordsListUid
}

export const selectLearnedWords = state =>
  state.toJS().learnedWords ? Array.from(state.toJS().learnedWords) : []

export const selectLearnedWordsInWordsList = state => {
  const wordsList = selectWordsList(state)
  return selectLearnedWords(state).filter(w => {
    return (
      Array.from(wordsList)
        .map(w => w.wordStr)
        .indexOf(w) !== -1
    )
  })
}

export const selectCheckinLogs = state =>
  state.toJS().checkinLogs
    ? Array.from(
        new Set(
          Array.from(state.toJS().checkinLogs)
            .filter(log => log)
            .map(log => log.substr(0, 8))
        )
      )
    : []

export const selectWordsListAmount = state => {
  const wordsList = selectWordsList(state)
  if (!wordsList) {
    return 0
  }
  return wordsList.length
}

export const selectDaysLeft = state => {
  const opts = selectOpts(state)
  const dailyWordsAmount = selectDailyWordsAmount(state)
  if (!opts) {
    return NaN
  }
  const daysLeft = models.StudyPlan.getDaysAmount(
    selectOpts(state),
    dailyWordsAmount
  )
  return daysLeft ? daysLeft : 1
}

export const selectShowDaysPicker = state => state.toJS().showDaysPicker

export const selectLearnedWordObjects = state => {
  const learnedWords = selectLearnedWords(state)
  return allWords.filter(word => {
    return learnedWords.indexOf(word.wordStr) !== -1
  })
}

export const selecteDailyGoalComplete = state => {
  return selectLearnedWordsOfCompletePractice(state) !== null
}

export const selectLearnedWordsOfCompletePractice = state => {
  const checkinLogs = state.toJS().checkinLogs
  const wordsListUid = state.toJS().wordsListUid
  const dateStringOfToday = new Date()
    .toISOString()
    .slice(0, 10)
    .replace(/-/g, '')
  const searchString = dateStringOfToday + '$' + wordsListUid
  const found = Array.from(checkinLogs)
    .filter(log => log)
    .filter(log => log.indexOf(searchString) !== -1)[0]

  if (!found) {
    return null // checkin log not found
  }

  let [date, listName, wordsAmount] = found.split('$')
  return Number.parseInt(wordsAmount)
}

export const selectWordsListComplete = state =>
  selectLearnedWordsInWordsList(state).length === selectWordsListAmount(state)
