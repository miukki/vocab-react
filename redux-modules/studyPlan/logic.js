/* global setInterval */

import { createLogic } from 'redux-logic'

import models from 'eic-vocab-app-models'
import { types as packageLogicTypes } from 'eic-vocab-app-redux-logic'

import types from './types'
import * as selector from './selector'
import wordsLists from '../../data/words_lists'
import StudyPlan from '../../lib/StudyPlan'

const checkDate = createLogic({
  type: types.CHECK_DATE,
  /* New option warnTimeout defaults to 60000 (ms == one minute) which warns (in development build only) when the logic exceeds the specified time without completion. Adjust this value or set it to 0 if you have logic that needs to exceed this time or purposefully never ends (like listening to a web socket) */
  warnTimeout: 0,
  process({ getState, action }, dispatch, done) {
    setInterval(() => {
      const date = new Date().toISOString().substr(0, 10)
      if (getState().toJS().studyPlan.currentDate !== date) {
        dispatch({ type: types.UPDATE_CURRENT_DATE })
      }
    }, 1000 * 60) // it check currentDate each 60 seconds to update state
  }
})

const loadSavedData = createLogic({
  type: types.LOAD_SAVED_DATA,
  process({ vocabAppServiceJsSdk, Storage, getState, action }, dispatch, done) {
    return StudyPlan.get({
      jwt: getState().toJS().token.jwt,
      vocabAppServiceJsSdk,
      Storage
    }).then(results => {
      if (results) {
        dispatch({ type: types.LOAD_SAVED_DATA_DONE, payload: results })
      }
      done()
    })
  }
})

const updateWordsList = createLogic({
  type: types.CHANGE_WORDS_LIST,
  processOptions: {
    dispatchReturn: true,
    successType: types.UPDATE_WORDS_LIST
  },
  process({ getState, action }) {
    return {
      wordsListUid: Array.from(wordsLists.keys())[action.payload.index]
    }
  }
})

const adjustDailyWordsPicker = createLogic({
  type: types.UPDATE_USER_DATA,
  processOptions: {
    dispatchReturn: true,
    successType: types.UPDATE_PICKER_DAILY_WORDS_AMOUNT
  },
  process({ getState, action }) {
    const dailyWordsAmount =
      action.payload.dailyWordsAmountOfWordsList[action.payload.wordsListUid] ||
      15
    return { dailyWordsAmount }
  }
})

const updateDailyWordsAmount = createLogic({
  type: types.CONFIRM_PICKER,
  processOptions: {
    dispatchReturn: true,
    successType: types.UPDATE_DAILY_WORDS_AMOUNT
  },
  process({ getState, action }) {
    return {
      dailyWordsAmount: getState().toJS().studyPlan.picker.dailyWordsAmount
    }
  }
})

const updateLearnedWords = createLogic({
  type: '@@vocab-app-redux-logic/practice-logic/word-learned', // for now
  // type: types.ADD_LEARNED_WORD,
  processOptions: {
    dispatchReturn: true,
    successType: types.UPDATE_LEARNED_WORDS
  },
  process({ getState, action }) {
    return {
      learnedWords: getState()
        .toJS()
        .studyPlan.learnedWords.add(action.wordStr) // for now
      // .studyPlan.learnedWords.add(action.payload.wordStr)
    }
  }
})

const updateCheckinLogs = createLogic({
  type: types.ADD_CHECKIN_LOG,
  processOptions: {
    dispatchReturn: true,
    successType: types.UPDATE_CHECKIN_LOGS
  },
  process({ getState, action }) {
    return {
      checkinLogs: getState()
        .toJS()
        .studyPlan.checkinLogs.add(action.payload.checkinLog)
    }
  }
})

const updateUserData = createLogic({
  type: [
    types.UPDATE_WORDS_LIST,
    types.UPDATE_DAILY_WORDS_AMOUNT,
    types.UPDATE_LEARNED_WORDS,
    types.UPDATE_CHECKIN_LOGS
  ],
  processOptions: {
    dispatchReturn: true,
    successType: types.UPDATE_USER_DATA
  },
  process({ getState, action }) {
    let {
      learnedWords,
      checkinLogs,
      wordsListUid,
      dailyWordsAmountOfWordsList
    } = getState().toJS().studyPlan

    if (action.payload.dailyWordsAmount) {
      dailyWordsAmountOfWordsList[wordsListUid] =
        action.payload.dailyWordsAmount
    }

    return {
      learnedWords,
      checkinLogs,
      wordsListUid,
      dailyWordsAmountOfWordsList,
      ...action.payload
    }
  }
})

const saveData = createLogic({
  type: types.UPDATE_USER_DATA,
  processOptions: {
    dispatchReturn: true,
    successType: types.SAVE_USER_DATA_DONE
  },
  process({ Storage, getState, action }) {
    return Storage.saveStudyPlan(action.payload)
  }
})

const renewPickerDailyWordsAmount = createLogic({
  // We are trying to keep previous value of dailyWordsAmount,
  // so this logic will not run when UPDATE_PICKER_DATA
  // but this may run in `checkOutRangePickerDailyWordsAmount` logic
  type: types.UPDATE_PICKER_DAYS_AMOUNT,
  process({ getState, action }, dispatch, done) {
    const opts = selector.selectOpts(getState().get('studyPlan'))

    const daysAmount = action.payload.daysAmount

    let dailyWordsAmount = models.StudyPlan.getDailyWordsAmount(
      opts,
      daysAmount
    )

    if (
      models.StudyPlan.getDaysAmount(
        opts,
        getState().toJS().studyPlan.picker.dailyWordsAmount
      ) !== daysAmount
    ) {
      dispatch({
        type: types.UPDATE_PICKER_DAILY_WORDS_AMOUNT,
        payload: { dailyWordsAmount }
      })
    }

    done()
  }
})

const renewPickerDaysAmount = createLogic({
  // We need to renew daysAmount valuse when
  // - picker data updated
  // - dailyWordsAmount changed
  type: types.UPDATE_PICKER_DAILY_WORDS_AMOUNT,
  process({ getState, action }, dispatch, done) {
    const opts = selector.selectOpts(getState().get('studyPlan'))

    const dailyWordsAmount = action.payload.dailyWordsAmount
    const daysAmount = models.StudyPlan.getDaysAmount(opts, dailyWordsAmount)

    if (getState().toJS().studyPlan.picker.daysAmount !== daysAmount) {
      dispatch({
        type: types.UPDATE_PICKER_DAYS_AMOUNT,
        payload: {
          daysAmount
        }
      })
    }

    done()
  }
})

const addCheckinLog = createLogic({
  type: packageLogicTypes.SHOW_QUESTION,
  process({ getState, action }, dispatch, done) {
    if (action.question === null) {
      dispatch({
        type: types.ADD_CHECKIN_LOG,
        payload: {
          checkinLog:
            new Date()
              .toISOString()
              .slice(0, 10)
              .replace(/-/g, '') +
            '$' +
            getState().toJS().studyPlan.wordsListUid +
            '$' +
            selector.selectDailyWordsAmount(getState().get('studyPlan'))
        }
      })
    }
    done()
  }
})

export default [
  checkDate,
  adjustDailyWordsPicker,
  loadSavedData,
  updateWordsList,
  updateDailyWordsAmount,
  updateLearnedWords,
  updateCheckinLogs,
  updateUserData,
  saveData,
  renewPickerDailyWordsAmount,
  renewPickerDaysAmount,
  addCheckinLog
]
