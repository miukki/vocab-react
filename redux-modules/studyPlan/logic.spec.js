import { fromJS } from 'immutable'
import { combineReducers } from 'redux-immutable'
import { createMockStore } from 'redux-logic-test'

import reducer, { initialState } from './'
import types from './types'
import logic from './logic'
import * as selector from './selector'
import { types as packageLogicTypes } from 'eic-vocab-app-redux-logic'

const rootReducer = combineReducers({ studyPlan: reducer })

describe('studyPlan logic', () => {
  describe('updateUserData', function() {
    it('dispatch UPDATE_USER_DATA with new dailyWordsAmountOfWordsList when dailyWordsAmount be specified', function(done) {
      selector.selectOpts = jest.fn()
      selector.selectOpts.mockImplementation(() => new Map([[40, 100]])) // to avoid crash

      const saveStudyPlan = jest.fn()
      saveStudyPlan.mockReturnValue(Promise.resolve())

      const store = createMockStore({
        initialState: fromJS({
          studyPlan: {
            checkinLogs: [],
            learnedWords: [],
            wordsListUid: 'IELTS Speaking',
            picker: { dailyWordsAmount: 15 },
            dailyWordsAmountOfWordsList: { 'IELTS Writing': 20 }
          }
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: {
          Storage: { saveStudyPlan }
        }
      })
      store.dispatch({
        type: types.UPDATE_DAILY_WORDS_AMOUNT,
        payload: {
          dailyWordsAmount: 40
        }
      })
      store.whenComplete().then(() => {
        expect(store.actions[1]).toMatchSnapshot()
        done()
      })
    })
  })

  describe('adjustDailyWordsPicker', function() {
    it('dispatch UPDATE_PICKER_DAILY_WORDS_AMOUNT with dailyWordsAmount of current wordsListUid', function(done) {
      selector.selectOpts = jest.fn()
      selector.selectOpts.mockImplementation(() => new Map([[20, 100]])) // to avoid crash

      const saveStudyPlan = jest.fn()
      saveStudyPlan.mockReturnValue(Promise.resolve())

      const store = createMockStore({
        initialState: fromJS({
          studyPlan: {
            picker: { dailyWordsAmount: 15 }
          }
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: {
          Storage: { saveStudyPlan }
        }
      })
      store.dispatch({
        type: types.UPDATE_USER_DATA,
        payload: {
          wordsListUid: 'IELTS Writing',
          dailyWordsAmountOfWordsList: { 'IELTS Writing': 20 }
        }
      })
      store.whenComplete().then(() => {
        expect(store.actions[1]).toMatchSnapshot()
        done()
      })
    })

    it('dispatch UPDATE_PICKER_DAILY_WORDS_AMOUNT with default value of dailyWordsAmount when dailyWordsAmount of current wordsListUid has not been set yet', function(done) {
      selector.selectOpts = jest.fn()
      selector.selectOpts.mockImplementation(() => new Map([[15, 100]])) // to avoid crash

      const saveStudyPlan = jest.fn()
      saveStudyPlan.mockReturnValue(Promise.resolve())

      const store = createMockStore({
        initialState: fromJS({
          studyPlan: {
            picker: { dailyWordsAmount: 15 }
          }
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: {
          Storage: { saveStudyPlan }
        }
      })
      store.dispatch({
        type: types.UPDATE_USER_DATA,
        payload: {
          wordsListUid: 'IELTS Speaking',
          dailyWordsAmountOfWordsList: { 'IELTS Writing': 20 }
        }
      })
      store.whenComplete().then(() => {
        expect(store.actions[1]).toMatchSnapshot()
        done()
      })
    })
  })

  describe('addCheckinLog', function() {
    it('dispatch ADD_CHECKIN_LOG', function(done) {
      Date.prototype.toISOString = jest.fn()
      Date.prototype.toISOString.mockReturnValue('2018-05-18')

      selector.selectOpts = jest.fn()
      selector.selectOpts.mockImplementation(() => new Map([[15, 100]])) // to avoid crash

      const saveStudyPlan = jest.fn()
      saveStudyPlan.mockReturnValue(Promise.resolve())

      const store = createMockStore({
        initialState: fromJS({
          studyPlan: {
            picker: {},
            wordsListUid: 'IELTS Listening',
            dailyWordsAmountOfWordsList: { 'IELTS Listening': 20 },
            checkinLogs: new Set(['20180501$IELTS Listening$15'])
          }
        }),
        reducer: rootReducer,
        logic,
        injectedDeps: {
          Storage: { saveStudyPlan }
        }
      })
      store.dispatch({
        type: packageLogicTypes.SHOW_QUESTION,
        question: null
      })
      store.whenComplete().then(() => {
        expect(store.actions[1]).toMatchSnapshot()
        expect(store.actions[2]).toMatchSnapshot()
        expect(store.actions[3]).toMatchSnapshot()
        done()
      })
    })
  })

  describe('checkDate', function() {
    // TODO: figure out how to test this
    // it('dispatch UPDATE_CURRENT_DATE', function(done) {
    //   Date.prototype.toISOString = jest.fn()
    //   Date.prototype.toISOString.mockReturnValue('2018-05-01')
    //   const store = createMockStore({
    //     initialState: fromJS({
    //       studyPlan: {
    //         currentDate: '2018-04-01'
    //       }
    //     }),
    //     reducer: rootReducer,
    //     logic,
    //     injectedDeps: {}
    //   })
    //   store.dispatch({
    //     type: types.CHECK_DATE
    //   })
    //   store.whenComplete().then(() => {
    //     expect(store.actions).toMatchSnapshot()
    //     done()
    //   })
    // })
  })
})
