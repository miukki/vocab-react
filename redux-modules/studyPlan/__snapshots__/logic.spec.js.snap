// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`studyPlan logic addCheckinLog dispatch ADD_CHECKIN_LOG 1`] = `
Object {
  "payload": Object {
    "checkinLog": "20180518$IELTS Listening$20",
  },
  "type": "eic-vocab-app/studyPlan/ADD_CHECKIN_LOG",
}
`;

exports[`studyPlan logic addCheckinLog dispatch ADD_CHECKIN_LOG 2`] = `
Object {
  "payload": Object {
    "checkinLogs": Set {
      "20180501$IELTS Listening$15",
      "20180518$IELTS Listening$20",
    },
  },
  "type": "eic-vocab-app/studyPlan/UPDATE_CHECKIN_LOGS",
}
`;

exports[`studyPlan logic addCheckinLog dispatch ADD_CHECKIN_LOG 3`] = `
Object {
  "payload": Object {
    "checkinLogs": Set {
      "20180501$IELTS Listening$15",
      "20180518$IELTS Listening$20",
    },
    "dailyWordsAmountOfWordsList": Object {
      "IELTS Listening": 20,
    },
    "learnedWords": undefined,
    "wordsListUid": "IELTS Listening",
  },
  "type": "eic-vocab-app/studyPlan/UPDATE_USER_DATA",
}
`;

exports[`studyPlan logic adjustDailyWordsPicker dispatch UPDATE_PICKER_DAILY_WORDS_AMOUNT with dailyWordsAmount of current wordsListUid 1`] = `
Object {
  "payload": Object {
    "dailyWordsAmount": 20,
  },
  "type": "eic-vocab-app/studyPlan/UPDATE_PICKER_DAILY_WORDS_AMOUNT",
}
`;

exports[`studyPlan logic adjustDailyWordsPicker dispatch UPDATE_PICKER_DAILY_WORDS_AMOUNT with default value of dailyWordsAmount when dailyWordsAmount of current wordsListUid has not been set yet 1`] = `
Object {
  "payload": Object {
    "dailyWordsAmount": 15,
  },
  "type": "eic-vocab-app/studyPlan/UPDATE_PICKER_DAILY_WORDS_AMOUNT",
}
`;

exports[`studyPlan logic updateUserData dispatch UPDATE_USER_DATA with new dailyWordsAmountOfWordsList when dailyWordsAmount be specified 1`] = `
Object {
  "payload": Object {
    "checkinLogs": Array [],
    "dailyWordsAmount": 40,
    "dailyWordsAmountOfWordsList": Object {
      "IELTS Speaking": 40,
      "IELTS Writing": 20,
    },
    "learnedWords": Array [],
    "wordsListUid": "IELTS Speaking",
  },
  "type": "eic-vocab-app/studyPlan/UPDATE_USER_DATA",
}
`;
